Initial etcd leader-election implementation.

venilnoronha opened the pull request:
Hi,

I've created a etcd implementation for spring-cloud-cluster leader-election along with autoconfiguration. I've also updated the tests and the leaderelection doc. Please pull.

pperalta commented on spring-cloud-cluster-etcd/src/main/java/org/springframework/cloud/cluster/etcd/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,327 @@
+		 */
+		private void notifyGranted() throws InterruptedException {
+			isLeader = true;
+			candidate.onGranted(context);
```
The contract for `onGranted` states:

> Callback method invoked when this candidate is elected leader. Implementations may chose to launch a background thread to perform leadership roles and return immediately. Another option is for implementations to perform all leadership work in the thread invoking this method. In the latter case, the method **must** respond to thread interrupts by throwing `java.lang.InterruptedException`. When the thread is interrupted, this indicates that this candidate is no longer leader.


If the implementation decides to perform its work in an endless loop, this will prevent `leaderEventPublisher` from publishing the application event (this is an issue with the other implementations; see https://github.com/spring-cloud/spring-cloud-cluster/issues/21). More critically, this will prevent the leader from renewing the TTL which will cause another process to become leader while this process continues to think that it is still the leader.

To honor the contract, I suggest the following:

1. Invoke `candidate.onGranted` in a separate thread. 
2. If leadership is revoked, that thread needs to be interrupted.

This is different from the ZooKeeper and Hazelcast implementations. The major difference is that with both of those, the underlying infrastructure maintains a connection with cluster peers and performs heartbeats; whereas with etcd that responsibility lies with the leader implementation.

Please let me know if you have any questions!

pperalta commented:
Hi Venil,

Thank you for your contribution! Have you signed the Individual Contributor Agreement? This is a requirement for us to accept pull requests from contributors:

https://support.springsource.com/spring_committer_signup

Thanks,
Patrick

venilnoronha commented:
Hi Patrick,

Thanks for the quick review. I've just signed the individual contributor agreement.

I'll try implementing the candidate.onGranted mechanism as you've suggested; will keep you updated on it.

venilnoronha commented:
Hi @pperalta,

Apart from the changes you had mentioned, I'm also spawning a separate thread for publishing revocation events. This is to avoid publishing of those revocation events before the GrantNotification execution actually finishes.

venilnoronha commented:
I've added 2 more tests.
1. leader yielding
2. revocation of leadership in blocking candidate implementations

venilnoronha commented:
I've added 2 more tests.
1. leader yielding
2. revocation of leadership in blocking candidate implementations

pperalta commented:
Hi Venil,

I had imagined a design with two threads:

### etcd leadership thread
* attempts to acquire leadership
* issues heartbeats
* manages worker thread
* raises leader events

### worker thread
* started by *etcd leadership* thread when leadership granted
* interrupted by *etcd leadership* thread when leadership revoked 

What do you think?

venilnoronha commented:
Hi Patrick,

In the design you've specified I would like to know how would leader yielding work?

As the worker thread itself would be the interrupting thread, we might not clearly understand when the worker thread would exit, if its a blocking candidate implementation. This is the reason I've kept another thread for interrupting the grantor and awaiting its termination.

venilnoronha commented:
I'll go through the designed you've proposed once again and let me see if I manage without the revoker thread.

Thanks,

Venil

pperalta commented:
Hi Venil,

I put together a quick prototype of what I was thinking and pushed it to here:

https://github.com/pperalta/spring-cloud-cluster/commit/55d1128e4b87d6ff5a8a44190bc1b20f85dc1551

Please have a look and let me know what you think.

Thanks,
Patrick

venilnoronha commented:
Hi Patrick,

Thank you so much for the example code. I've implemented it accordingly now.

Regards,
Venil

venilnoronha commented:
Hi Patrick,

Were you able to verify the final implementation?

pperalta commented:
Hi Venil,

I am still working on reviewing your changes. We had a long holiday weekend in the US so I haven't looked in a few days. I'll update the PR with comments soon!

venilnoronha commented:
OK.

Thanks Patrick.

pperalta commented on spring-cloud-cluster-etcd/src/main/java/org/springframework/cloud/cluster/etcd/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,351 @@
+		stop();
+		workerExecutorService.shutdown();
+		leaderExecutorService.shutdown();
+		workerExecutorService.awaitTermination(Long.MAX_VALUE, TimeUnit.DAYS);
```
Why do we need to block the thread until the service has been shut down?

venilnoronha replied:
@pperalta If the current node was the cluster leader, Initiator would try to delete the candidate key on etcd (via the 'tryDeleteCandidateEntry' method) while shutting down. Now, EtcdClient which issues requests via its internal thread pool also implements the Closable interface, to shut down the said thread pool. When the Spring container is shutting down, EtcdClient's 'close' method gets called, causing it to accept no more requests i.e. if EtcdClient is closed before calling the 'tryDeleteCandidateEntry' method, its etcd request would fail. So, to avoid the etcd call failure, I'm blocking the shutdown thread until the Initiator and Worker threads finish execution i.e. the EtcdClient's 'close' method execution is delayed until these threads exit.

Let me know what you think or if you have any alternate ideas.

Thanks,

Venil

pperalta replied:
That makes sense. The only thing I would change is to set the timeout to something reasonable (perhaps 30 seconds) so that the thread shutting down the application context doesn't get stuck forever if one of the executor threads gets stuck.

pperalta commented on spring-cloud-cluster-etcd/src/main/java/org/springframework/cloud/cluster/etcd/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,351 @@
+				// If the candidate's leadership was revoked
+			}
+			finally {
+				candidate.onRevoked(context);
```
If a runtime exception is thrown by `candidate.onGranted()`, `candidate.onRevoked()` will be invoked - however it appears that the leadership thread will continue issuing heartbeats even though the candidate has relinquished leadership.

venilnoronha replied:
I've fixed this issue with commit 05aca24b037d89f31921e0c4c4e6f9c193a8f442. I've added a test for it too.

pperalta commented on spring-cloud-cluster-etcd/src/main/java/org/springframework/cloud/cluster/etcd/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,351 @@
+						TimeUnit.SECONDS.sleep(HEART_BEAT_SLEEP);
+					}
+					catch (IOException | TimeoutException e) {
+						LoggerFactory.getLogger(getClass()).warn("Exception occurred while trying to access etcd", e);
```
If an exception is thrown by `sendHeartBeat`, should leadership be relinquished?

venilnoronha replied:
There can be 2 scenarios in that case,

1. Etcd is not accessible only by the current leader.
Impact: If we don't relinquish leadership, some other node would become leader simultaneously.

2. Etcd is not accessible by any of the cluster components.
Impact: If we relinquish the current node's leadership, then there might be no cluster leader at all.

How should we really handle this situation? The second option sounds good, but, having no leader at all might be dangerous for certain types of components.

pperalta replied:
In general, I think it is safer for leadership to be relinquished when the leader is not able to issue a heartbeat. The idea behind electing a leader in a distributed system is that there is only ever at most one leader.

This is how error handling is implemented with Curator's leadership election recipe:

http://curator.apache.org/curator-recipes/leader-election.html

> The recommended action for receiving SUSPENDED or LOST is to throw `CancelLeadershipException`. This will cause the `LeaderSelector` instance to attempt to interrupt and cancel the thread that is executing the `takeLeadership` method.

With ZooKeeper and Hazelcast, the underlying library issues heartbeats to ensure connectivity to the cluster.  If that heartbeat cannot be maintained, an error is thrown and leadership is relinquished. Since the etcd leadership implementation issues its own heartbeat, it is responsible for relinquishing leadership when it cannot maintain the heartbeat.

venilnoronha replied:
Sure. I'll work on this and update you.

Thanks,

Venil

venilnoronha replied:
I've fixed this with e15f330.

venilnoronha commented:
Hi Patrick,

Could you please go through these commits; I've tried to fix the problems you've highlighted.

Thanks,
Venil

pperalta commented on spring-cloud-cluster-etcd/src/main/java/org/springframework/cloud/cluster/etcd/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,377 @@
+	class Worker implements Callable<Void> {
+
+		@Override
+		public Void call() {
```
I think this can be simplified to:

```
@Override
public Void call() throws InterruptedException {
	try {
		candidate.onGranted(context);
		Thread.sleep(Long.MAX_VALUE);
	}
	finally {
		relinquishLeadership = true;
		candidate.onRevoked(context);
	}
	return null;
}
```

pperalta commented on spring-cloud-cluster-etcd/src/main/java/org/springframework/cloud/cluster/etcd/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,377 @@
+				while (running) {
+					if (relinquishLeadership) {
+						relinquishLeadership = false;
+						relinquishLeadership();
```
I would reset the `relinquishLeadership` flag after `relinquishLeadership()` is invoked. Consider the following possibility:

Worker thread: `candidate.onGranted` throws an Exception
Worker thread: `relinquishLeadership` set to `true`

Leader thread: `relinquishLeadership` set to `false`
Leader thread: invoke `relinquishLeadership()`

Other thread: invoke `context.yield()`
Other thread: `relinquishLeadership` set to `true` while Leader thread is invoking  `relinquishLeadership()`

Leader thread: sees that `relinquishLeadership == true`, so it invokes `relinquishLeadership()` a second time, which causes a second event to be published to `leaderEventPublisher`.

pperalta commented:
I pushed some suggestions to https://github.com/pperalta/spring-cloud-cluster/commit/bb6676c95e6b870570439b171a62dbfd7c0569e5 for your consideration. It includes changes based on my latest feedback, and it attempts to simplify exception handling.

I was also thinking that this project would be an interesting candidate to use the new [Spring State Machine](http://projects.spring.io/spring-statemachine/) project. Since we are tracking multiple states with various `if` `then` `else` statements, using a state machine could simplify the implementation. This is your PR so there is no obligation, but I thought that I'd make the suggestion in case you're interested. 😊

venilnoronha commented:
Hi Patrick,

I've updated the implementation as per your changes. Also, I've modified the etcd configuration properties to accept connect uris in a comma separated fashion instead of a YAML list one.

I'm interested in implementing this feature using Spring StateMachine that you mentioned; I'm planning to create a separate PR for it. Should that be fine?

pperalta commented:
Yes, if you want to try with the state machine that would be great! In the meantime do you want me to merge this PR or do you want me to wait until you get a chance to try the state machine?

venilnoronha commented:
I would request you to please merge the PR while I work on the state machine base implementation.

Thanks Patrick for bearing with me :)

Regards,
Venil

pperalta commented:
Rebased, squashed, and merged as ed17e15dc823d37059800022fa843c251ca902c5. Thanks Venil!
